contracts/SpeedMarkets/ChainedSpeedMarketsAMM.sol

```diff
@@ -0,0 +1,521 @@
+        _resolveMarket(market, priceUpdateData);
+        uint amountDiff = sUSD.balanceOf(user) - amountBefore;
+        sUSD.safeTransferFrom(user, address(this), amountDiff);
+        if (amountDiff > 0) {
+            if (toEth) {
+                uint offramped = multiCollateralOnOffRamp.offrampIntoEth(amountDiff);
+                address payable _to = payable(user);
+                bool sent = _to.send(offramped);
+                require(sent, "Failed to send Ether");
+            } else {
+                uint offramped = multiCollateralOnOffRamp.offramp(collateral, amountDiff);
+                IERC20Upgradeable(collateral).safeTransfer(user, offramped);
+            }
+        }
+    }
+
+    /// @notice resolveMarkets in a batch
+    function resolveMarketsBatch(address[] calldata markets, bytes[][] calldata priceUpdateData)
+        external
+        payable
+        nonReentrant
+        notPaused
+    {
+        for (uint i = 0; i < markets.length; i++) {
+            if (canResolveMarket(markets[i])) {
+                bytes[][] memory subarray = new bytes[][](1);
+                subarray[0] = priceUpdateData[i];
+                _resolveMarket(markets[i], subarray);
+            }
+        }
+    }
+
+    function _resolveMarket(address market, bytes[][] memory priceUpdateData) internal {
+        require(canResolveMarket(market), "Can not resolve");
+
+        bytes32[] memory priceIds = new bytes32[](1);
+        priceIds[0] = speedMarketsAMM.assetToPythId(ChainedSpeedMarket(market).asset());
+
+        int64[] memory prices = new int64[](priceUpdateData.length);
+        uint64 strikeTimePerDirection;
+        for (uint i = 0; i < priceUpdateData.length; i++) {
+            strikeTimePerDirection =
+                ChainedSpeedMarket(market).initialStrikeTime() +
+                uint64(i * ChainedSpeedMarket(market).timeFrame());
+
+            PythStructs.PriceFeed[] memory pricesPerDirection = speedMarketsAMM.pyth().parsePriceFeedUpdates{
+                value: speedMarketsAMM.pyth().getUpdateFee(priceUpdateData[i])
+            }(
+                priceUpdateData[i],
+                priceIds,
+                strikeTimePerDirection,
+                strikeTimePerDirection + speedMarketsAMM.maximumPriceDelayForResolving()
+            );
+
+            PythStructs.Price memory price = pricesPerDirection[0].price;
+            require(price.price > 0, "invalid price");
+            prices[i] = price.price;
+        }
+
+        _resolveMarketWithPrices(market, prices);
+    }
+
+    /// @notice admin resolve market for a given market address with finalPrice
+    function resolveMarketManually(address _market, int64[] calldata _finalPrices) external isAddressWhitelisted {
+        _resolveMarketManually(_market, _finalPrices);
+    }
+
+    /// @notice admin resolve for a given markets with finalPrices
+    function resolveMarketManuallyBatch(address[] calldata markets, int64[][] calldata finalPrices)
```

Comment:
as we are going to need a hot wallet for resolving for capital efficiency, we need to make sure these markets that are manually resolved are all in favor of AMM. Same goes for Single Speed Markets.